src/daft-micropartition/src/table_stats/mod.rs

```diff
@@ -0,0 +1,143 @@
+use std::{collections::HashSet, ops::Not};
+
+use daft_dsl::Expr;
+use daft_table::Table;
+use indexmap::IndexMap;
+use snafu::ResultExt;
+
+use crate::column_stats::{self, ColumnRangeStatistics};
+
+use daft_core::array::ops::DaftCompare;
+
+#[derive(Clone, Debug)]
+pub(crate) struct TableStatistics {
+    pub columns: IndexMap<String, ColumnRangeStatistics>,
+}
+impl TableStatistics {
+    fn from_table(table: &Table) -> Self {
+        let mut columns = IndexMap::with_capacity(table.num_columns());
+        for name in table.column_names() {
+            let col = table.get_column(&name).unwrap();
+            let stats = ColumnRangeStatistics::from_series(col);
+            columns.insert(name, stats);
+        }
+        TableStatistics { columns }
+    }
+}
+
+impl TableStatistics {
+    pub(crate) fn union(&self, other: &Self) -> crate::Result<Self> {
+
+        // maybe use the schema from micropartition instead
+        let unioned_columns = self
+            .columns
+            .keys()
+            .chain(other.columns.keys())
+            .collect::<HashSet<_>>();
+        let mut columns = IndexMap::with_capacity(unioned_columns.len());
+        for col in unioned_columns {
+            let res_col = match (self.columns.get(col), other.columns.get(col)) {
+                (None, None) => panic!("Key missing from both tables; invalid state"),
+                (Some(_l), None) => Ok(ColumnRangeStatistics::Missing),
+                (None, Some(_r)) => Ok(ColumnRangeStatistics::Missing),
+                (Some(l), Some(r)) => l.union(r),
+            }?;
+            columns.insert(col.clone(), res_col);
+        }
+        Ok(TableStatistics { columns })
+    }
+
+    pub(crate) fn eval_expression(&self, expr: &Expr) -> crate::Result<ColumnRangeStatistics> {
+        match expr {
+            Expr::Alias(col, _) => self.eval_expression(col.as_ref()),
+            Expr::Column(col) => {
+                let col = self.columns.get(col.as_ref()).unwrap();
+                Ok(col.clone())
+            }
+            Expr::Literal(lit_value) => lit_value.try_into(),
+            Expr::Not(col) => self.eval_expression(col)?.not(),
+            Expr::BinaryOp { op, left, right } => {
+                let lhs = self.eval_expression(left)?;
+                let rhs = self.eval_expression(right)?;
+                use daft_dsl::Operator::*;
+                match op {
+                    Lt => lhs.lt(&rhs),
+                    LtEq => lhs.lte(&rhs),
+                    Eq => lhs.equal(&rhs),
+                    NotEq => lhs.not_equal(&rhs),
+                    GtEq => lhs.gte(&rhs),
+                    Gt => lhs.gt(&rhs),
+                    Plus => &lhs + &rhs,
+                    Minus => &lhs - &rhs,
+                    _ => todo!(),
```

Comment:
Should we return a `Missing` here instead of todo? Ditto for the todo at the bottom.